posts/inside-rust/2023-12-22-trait-system-refactor-initiative.md

```diff
@@ -0,0 +1,57 @@
+
+## Non-fatal overflow
+
+Unlike the existing trait solver, the new solver does not immediately abort compilation when hitting the recursion limit:
+```rust
+struct Wrapper<T>(T);
+
+trait Overflow {}
+impl<T> Overflow for Wrapper<Wrapper<T>> where Wrapper<T>: Overflow {}
+impl Overflow for Wrapper<u32> {}
+
+// Checking whether these two implementations overlap
+// tries to prove that either `Wrapper<_>: Overflow` or
+// `Wrapper<_>: Copy` do not hold.
+//
+// The existing solver first checks `Wrapper<_>: Overflow`,
+// resulting in overflow and aborting compilation.
+//
+// The new solver does not abort compilation on overflow and
+// considers the implementations to be disjoint, given that
+// `Wrapper<_>: Copy` does not hold.
+trait MayOverlap {}
+impl<T: Overflow + Copy> MayOverlap for T {}
+impl<T> MayOverlap for Wrapper<T> {}
+```
+
+This change is necessary as popular crates, e.g. [typenum](https://github.com/rust-lang/trait-system-refactor-initiative/issues/73), now reach the recursion limit with the new solver and would therefore break if overflow remains fatal. This is caused by [the removal of a heuristic present in the old solver](https://github.com/rust-lang/trait-system-refactor-initiative/issues/56). It is also desirable as the compilation result is otherwise order dependent. This order dependence complicates future changes to the type system, e.g. [an attempt to switch to deferred projection equality in the old solver](https://github.com/rust-lang/rust/pull/96912) also ended up causing an overflow error in typenum, preventing it from getting merged. It is also observable by users, e.g. switching the order of where-clauses in the above example to `T: Copy + Overflow` causes this snippet to compile with the old solver.
+
+The new solver now returns overflow when hitting the recursion limit. However, this change by itself causes the solver to very easily hang due to exponential blowup. We therefore greatly limit the available recursion depth for later goals after encountering overflow and also discard some of the inference constraints from goals resulting in overflow.
+
+It is important to balance the performance of the trait solver with the expressiveness of the trait system. While we're fairly happy with the approach we've settled on for now, getting this right is more art than science. We believe our current approach to be performant in most cases and to allow for further significant performance optimizations in the future. We also expect it to provide the necessary expressiveness to be backwards compatible with the old solver and to behave as expected by users.
```

Comment:
```suggestion
It is important to balance the performance of the trait solver with the expressiveness of the trait system. While we're fairly happy with the approach we've settled on for now, getting this right is more art than science. We believe our current approach to be performant in most cases and to allow for further significant performance optimizations in the future. We also expect it to provide the necessary expressiveness to be backward compatible with the old solver and to behave as expected by users.
```

Reply:
both are viable, I prefer using "backwards"

Reply:
Specifically in the case of the phrase "backwards compatible"? because I dont think ive ever heard anyone say it like "backward compatible" lol

Reply:
Yes, "backwards compatible" feels more idiomatic to me, even though they're in free variation I guess

Reply:
Wikipedia omits the S: https://en.wikipedia.org/wiki/Backward_compatibility
Merriam-Webster: https://www.merriam-webster.com/dictionary/backward%20compatible
And Wiktionary: https://en.wiktionary.org/wiki/backward_compatible

All 3 list with-s as a less-common variant though